### PR TITLE
bpo-31976: Fix race condition when flushing a file is slow.

### PR DESCRIPTION
**Pull request by benfogle**
* Move IS_CLOSED check to after the acquiring the lock
* Add a check for self->buffer == NULL in IS_CLOSED
* Make _pyio.BufferedWriter close the same way the C version does
* Add unit tests

https://bugs.python.org/issue31976

### CODE REVIEW

File: Lib/_pyio.py
Comment (pitrou) [inline Lib/_pyio.py]: This doesn't look right. The `close()` method should take the lock once, not take it three times in a row (one of them implicit in `flush()`). Note `_flush_unlocked()` can help you here.

File: Modules/_io/bufferedio.c
Comment (pitrou) [inline Modules/_io/bufferedio.c]: Can you add a comment and a reference to the bug entry here?

File: Lib/test/test_io.py
Comment (pitrou) [inline Lib/test/test_io.py]: Can you add a comment pointing to the bug entry here?

File: Lib/test/test_io.py
Comment (pitrou) [inline Lib/test/test_io.py]: You could probably use an `Event` that you would set in `SlowFlushRawIO.flush()` instead.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for spotting this and proposing this fix!
Before this goes in, you should address the comments and also a NEWS entry. For the latter, you can use [the "blurb" utility](https://pypi.python.org/pypi/blurb).

File: Lib/_pyio.py
Comment (benfogle) [reply Lib/_pyio.py]: That made a bunch of tests (`test_flush_error_on_close`, etc.) fail since `close()` no longer calls `flush()` if the file object is subclassed or `fileobj.flush` is set to something. 

This is actually the same behavior as `buffered_close` in the C implementation. I have added a comment to that effect.

**Comment by benfogle** [comment]
I have made the requested changes; please review again

File: Lib/test/test_io.py
Comment (pitrou) [reply Lib/test/test_io.py]: It looks like you forgot to address this comment.

Comment (benfogle) [reply Lib/test/test_io.py]: Ah, whoops. Thank you.

**Comment by benfogle** [comment]
I have made the requested changes; please review again.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thanks. I'm inclined to merge this after CI approval. I'm not sure I want to backport it to 3.6, though, because there is a tiny but non-zero chance that this might break some odd usage somewhere and I don't want that to happen in a bugfix release.

**Pull request merged**
